source/octf/trace/parser/ParsedIoTraceEventHandler.cpp

```diff
@@ -119,38 +291,76 @@ ParsedIoTraceEventHandler::getEventMessagePrototype() {
 }
 
 void ParsedIoTraceEventHandler::flushEvents() {
-    auto iter = m_cache.begin();
-    auto end = m_cache.end();
-    for (; iter != end;) {
-        if (iter->second.mutable_io()->latency()) {
-            handleIO(iter->second);
-            iter = m_cache.erase(iter);
+    while (m_queue.size()) {
+        auto &event = m_queue.front();
+        if (event.io().latency()) {
+            pushOutEvent(event);
+            m_queue.pop();
         } else {
             break;
         }
     }
 
     bool isFinished = getParser()->isFinished();
 
-    constexpr uint64_t cacheLimit = 1000 * 1000;
-    if (m_cache.size() < cacheLimit && !isFinished) {
+    if (m_queue.size() < m_limit && !isFinished) {
         return;
     }
 
     // Cache exceeds some number,
     // or every parser finished its job, then flush IOs
-    while (m_cache.size()) {
-        handleIO(m_cache.begin()->second);
-        m_cache.erase(m_cache.begin());
+    while (m_queue.size()) {
+        auto &event = m_queue.front();
+        pushOutEvent(event);
+        m_queue.pop();
 
-        if (m_ioQueueDepth) {
-            m_ioQueueDepth--;
-        }
-
-        if (!isFinished && m_cache.size() < cacheLimit) {
+        if (!isFinished && m_queue.size() < m_limit) {
             break;
         }
     }
 }
 
+void ParsedIoTraceEventHandler::pushOutEvent(proto::trace::ParsedEvent &event) {
+    // Get SID
+    auto sid = event.header().sid();
+
+    // Remove SID mapping
+    m_sidMapping.erase(sid);
+
+    // Remove LBA mapping
+    Key key(event);
+    m_lbaMapping->erase(key, sid);
+
+    // Update SID
+    event.mutable_header()->set_sid(sid - m_sidOffset);
+
+    // Take into account IO queue depth adjustment
+    auto devId = event.device().id();
+    auto &qd = m_devIoQueueDepth[devId];
+    auto ioqd = event.io().qd();
+    if (qd.Adjustment < ioqd) {
+        ioqd -= qd.Adjustment;
```

Comment:
Why is Adjustment needed as a separate field?

Reply:
It's because we want to avoid updating already queued events. Before pushing out events we consider adjustment.